### DISCUSSION TITLE
Support S3 API for JetStream

### DISCUSSION THREAD

**Discussion by khalil-omer** [nats-io/nats-server#5486] [Ideas]
JetStream requires a fast local block storage service. This is great for use cases like KV and for read heavy workflows. However, there are a number of products that are using S3 as the fundamental durability layer and treating disks as just caches, and this pattern could have value for a lot of NATS use cases. 

Examples of products using this pattern are WarpStream (Kafka on S3), Neon (Postgres on S3), and Axiom (log analytics).  

It could be useful to explore how NATS could potentially benefit from supporting this as a configuration option. 

I imagine the basic flow as follows:

1. `publish()` to NATS server as usual (no change to client libraries)
2. The JetStream instance that received the message tries to write it to S3. If it succeeds, the publish is ack'd. (Batching could be enabled by forcing clients to wait for up to 200ms)
3. The JetStream instance now pushes this write to all other nodes in the cluster like normal so that they can publish it to subscribers/consumers. What's different is that none of the nodes are responsible for persisting the message.

This would be a pretty involved project, but it would unlock infinite and low cost stream scale.

**Comment by derekcollison** [comment]
We have some thoughts around virtual blocks, which is an approach that could use S3 to offload these when not needed locally.

Timeframe would be 2.12 though..

**Comment by hpvd** [comment]
see also S3 next level: offload, backup plus also data-sharing, querying? https://github.com/nats-io/nats-server/discussions/6478

**Comment by sethitow** [comment]
I've implemented a JetStream storage backend that uses S3-compatible object storage. This implementation is not suitable for production use, but hopefully someone might find it interesting. The patch can be [viewed here](https://github.com/nats-io/nats-server/compare/main...sethitow:nats-server:objstore?expand=1). A blog post with more information and design discussion is [here](https://www.sethitow.com/posts/2025-nats-jetstream-storage-backend-s3-object-storage/).

**Reply by derekcollison** [reply]
Thanks Seth.

**Reply by PhE** [reply]
Great ! Hope it could evolve  into a supported jetstream backend

**Comment by BasixKOR** [comment]
For reference, Kafka is seeking to incorporate Object Storage based storage as outlined by [KIP-1150](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1150%3A+Diskless+Topics) and other related proposals.

**Comment by joeblew999** [comment]
Would be crazy awesome if this was real 

using s3 and then caching to fs like how Duckdb and other thing do now makes a lot of infrastructure headaches disappear. S3 disk is so much cheaper then dedicated disk too !
